### PR TITLE
UX - Expandable Search Patient By Any Demographics button with mobile…

### PR DESCRIPTION
**Pull request by zbig01**
… friendly Patient Finder

Combined the Search Patient button in main.php with the Any Demographics text box
to create an expandable search box that searches for patient by part or whole of any demographics field

Works on all major modern browsers and mobile devices and styled for rtl languages as well

Patient Finder is mobile friendly up to iPhone 6 landscape mode

Files changed:

openemr/interface/main/finder/dynamic_finder.php
openemr/interface/main/tabs/js/user_data_view_model.js
openemr/interface/main/tabs/templates/patient_data_template.php
openemr/interface/main/tabs/main.php
openemr/interface/themes/rtl.scss
openemr/interface/themes/oe-common/main-common.scss
openemr/interface/themes/style_light.scss
openemr/interface/themes/style_manila.scss
openemr/interface/themes/buttons/_manilla.scss
openemr/src/OeUI/OemrUI.php

Don't forget to run `npm run build` :)

**Collapsed Expandable Search button**

expandable_search_01

**Expanded Expandable Search button**

expandable_search_02

**Patient Finder - Desktop condensed view**

expandable_search_03

**Patient Finder - Desktop expanded view**

expandable_search_04

**Patient Finder - iPad - Portrait View** 

expandable_search_06

**Patient Finder - iPad - Landscape View** 

expandable_search_07

**Patient Finder - iPhone 6.7.8 - Landscape View - note no superfluous icons**

expandable_search_08

**Patient Finder - iPhone 6.7.8 - Landscape View - search box**

expandable_search_09

**Patient Finder - iPhone 6.7.8 - Portrait View -note table with minimal horizontal scrolling**

expandable_search_10

below added by bradymiller:
Up For Grabs demo link to test this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Zeta_-_Up_For_Grabs_Demo

### CODE REVIEW

File: interface/main/finder/dynamic_finder.php [hunk 1]
```diff
@@ -162,33 +170,39 @@ function persistCriteria(el, e) {
...
+                    <?php echo $oemr_ui->pageHeading() . "\r\n"; ?>
+                </div>
             </div>
-            <div class="col-sm-2">
+        </div>
+        <div class="row">
+            <div class="col-sm-12">
                 <?php if (acl_check('patients', 'demo', '', array('write','addonly'))) { ?>
-                    <button id="create_patient_btn" onclick="top.restoreSession();top.RTop.location = '<?php echo $web_root ?>/interface/new/new.php'"><?php echo xlt('Create Patient'); ?></button>
+                    <button id="create_patient_btn1" class="btn btn-default btn-add" onclick="top.restoreSession();top.RTop.location = '/openemr-master/interface/new/new.php'">Add New Patient</button>
```
Comment (bradymiller) [inline interface/main/finder/dynamic_finder.php]: Guessing the above line change is not meant to be there ( hard-codes path, doesn't translate)

File: interface/main/finder/dynamic_finder.php [hunk 2]
```diff
@@ -162,33 +170,39 @@ function persistCriteria(el, e) {
...
                 <?php } ?>
+                <i id="search_hide" class="fa fa-search fa-lg oe-pull-away" aria-hidden="true"></i>
             </div>
+            <!--<div class="col-sm-2 oe-pull-away">
+                <i id="search_hide" class="fa fa-search fa-lg" aria-hidden="true"></i>
+            </div>-->
```
Comment (bradymiller) [inline interface/main/finder/dynamic_finder.php]: better to not include commented out code

**Comment by bradymiller** [comment]
Up For Grabs demo link to test this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Zeta_-_Up_For_Grabs_Demo

File: interface/main/finder/dynamic_finder.php (hunk 1 above)
Comment (zbig01) [reply interface/main/finder/dynamic_finder.php]: Sorry missed in cleanup

File: interface/main/finder/dynamic_finder.php (hunk 2 above)
Comment (zbig01) [reply interface/main/finder/dynamic_finder.php]: sorry missed in cleanup

**Comment by zbig01** [comment]
Made changes, the demo is working well. If you can pl test on a newer iPad

**Comment by sjpadgett** [comment]
The any search needs to default to open. The point for the search as it was is for quick easy access.

**Comment by bradymiller** [comment]
I'm getting odd behavior on this when it is open. When I click the search icon, sometimes it opens the search tab and other times is just closes the search bar thing.

**Comment by sjpadgett** [comment]
I'm kind of okay with the search expand because it only takes one click still to start a search however, the primary purpose of the search icon is for a quick way to open finder. I think it should be kept static unless there is a reason for it to be collapsed.

**Comment by zbig01** [comment]
@sjpadgett
There are several reasons why the Search by any Demographics feature was re-styled this way:
    First and foremost the top strip that primarily exists to hold patient data has only action 'buttons' in that strip, i.e. the user has to click on any of the 'buttons' to do something. Thus keeping the Search by any Demographics hidden and displaying only the search button conforms with  all the other action elements on that strip.
    
    The Expandable Search by any demographics button emphasizes this behavior i.e. clicking and brings about a uniformity to actions that can be performed by the user. However users are used to pressing the 'Enter' key to submit a form and this functionality that you had designed was retained in a little jQuery script on the main.php page.
    
    While real estate is not a huge problem on a desktop it becomes progressively greater as we scale down, an open text box consumes more and more of the display as we access it on smaller devices. To keep the design responsive and to ensure that all the elements are balanced across displays it is imperative that the text box be hidden.
    
    Furthermore keeping the text box hidden frees us to use a larger text box which is easier to use and where its placeholder text can be fully displayed to provide a meaningful hint to the user.

**Comment by zbig01** [comment]
@sjpadgett 

There are several reasons why the Search by any Demographics feature was re-styled this way:
First and foremost the top strip that primarily exists to hold patient data has only action 'buttons' in that strip, i.e. the user has to click on any of the 'buttons' to do something. Thus keeping the Search by any Demographics hidden and displaying only the search button conforms with  all the other action elements on that strip.

The Expandable Search by any demographics button emphasizes this behavior i.e. clicking and brings about a uniformity to actions that can be performed by the user. However users are used to pressing the 'Enter' key to submit a form and this functionality that you had designed was retained in a little jQuery script on the main.php page.

While real estate is not a huge problem on a desktop it becomes progressively greater as we scale down, an open text box consumes more and more of the display as we access it on smaller devices. To keep the design responsive and to ensure that all the elements are balanced across displays it is imperative that the text box be hidden.

Furthermore keeping the text box hidden frees us to use a larger text box which is easier to use and where its placeholder text can be fully displayed to provide a meaningful hint to the user.

**Comment by sjpadgett** [comment]
I don't want to sacrifice functionality for the sake of trying to run OpenEMR on phones. The search icon still needs to either initiate a search or open finder. The text box is contained within the patient info/photo group and should be responsive as a group ie the same real estate is needed whether collapsed or not.
Our users rely on that icon and changing behavior to accommodate collapsing a search box should be avoided. There is nothing unusual at all having a search box in any header.

**Comment by sjpadgett** [comment]
I'm not trying to give ya a hard time but I see this will be an issue after release.

**Comment by zbig01** [comment]
@bradymiller 

The odd behavior where at times the Search Patient button would just close the Search Any text box when it was clicked on the `main.php` patient section strip was due to the following reason:

It was bound to the `click` event like the old Search button.  
  
 `data-bind="click: viewPtFinder.bind( $data, '<?php echo xla("The search field cannot be empty. Please enter a search term") ?>')"`
 
 This triggered the `function viewPtFinder()` in `/interface/main/tabs/js/user_data_view_model.js`. Unlike the old `function viewPtFinder() `the new one does a few extra things, the first of which is to `set focus` to the search any text box causing it to expand and be displayed via the css rule` .oe-expandable-search input:focus`. 
 
 The problem with that binding was that the `mousedown` event that is the first part of the `click` event causes the focus to move from the input box to the search button itself, especially if you completed the `click `by executing the `mouseup` after some delay. Upon losing focus the expected behavior for the input box is to revert to its default collapsed state. 

However the code block in` function viewPtFinder() `to open `/interface/main/finder/dynamic_finder.php` i.e. the Patient Finder page
   ```
 navigateTab(finderUrl,"fin", function () {
        activateTabByName("fin",true);
    }); 
```
can only be executed on an open input text box that contains a non empty string. Thus failing to achieve the intended effect.

If you completed the `click` by quickly releasing the mouse you will notice the Search Patient button starts to close, but as the` function viewPtFinder()` gets executed by the now completed `click `event the focus moves back to the text box and it stays open allowing the rest of the code to be executed.
 
By binding the Search Patient button to the `mousedown` event instead this problem has been eliminated. You can now click as slowly as you want.
 
`data-bind="event: {mousedown: viewPtFinder.bind( $data, '<?php echo xla("The search field cannot be empty. Please enter a search term") ?>')}"`

So far in my testing this seems to have fixed the problem

**Comment by zbig01** [comment]
@bradymiller 
The PR #2513 fixes above issues, current PR may be closed :)

**Comment by bradymiller** [comment]
hi @zbig01 , will get a up for grabs demo going with the new PR

**Pull request closed without merging**